bgpd/bgp_route.c

```diff
@@ -1800,6 +1800,15 @@ int subgroup_announce_check(struct bgp_node *rn, struct bgp_path_info *pi,
 		if (!bgp_outbound_policy_exists(peer, filter))
 			return 0;
 
+	/* Do not send prefixes with nexthop address as peer. */
```

Comment:
What problem are you trying to solve? What happens with nexthop self? If you want to change such a fundamental default behavior, you should put this under a cli knob that you can enable.

Reply:
```
A route originated by a BGP speaker SHALL NOT be advertised to a peer
   using an address of that peer as NEXT_HOP.  A BGP speaker SHALL NOT
   install a route with itself as the next hop.
```

Reply:
https://www.ietf.org/rfc/rfc4271.txt

Reply:
Appreciate the pointer. I am aware of what the RFC says. My questions are somewhat different though but I will rephrase.

1) What problem is this solving for you?
2) What happen if nexthop self is set? Aren't you bailing out too early?

Also please note the RFC says SHALL and not MUST.

Reply:
Basically not a problem at all, IMO. 

Otherwise, why should we flood the network with unnecessary data if we can avoid that?

Reply:
Good question. Here are the reasons from my point of view.

1) Such a change can easily break things (I think this was demonstrated already).
2) Sending the routes back is much cheaper than preventing them and the receiver has many ways to detect and reject the routes during incoming update processing which is way less expensive than update generation in BGP.
3) Have you performed any analysis on what is the impact of this change in update generation/ update groups? Think of the example where the peer that sent the update, is in the same group with other peers.
4) To minimize the risk of breaking functionality (since we're not trying to fix something) by changing established default behavior, we should have this change under a knob.

Reply:
1. It was (and it is) because I don't know how to deny route advertisement per peer and not per peer-group. Because at the moment I can filter by peer-group.

2. Basically, the receiver side drops prefixes with next-hop implicitly.

3. This is what I want to figure out, but still, do not have much experience in how this works, as mentioned in number one :point_up:  :)

I agree with you in all aspects, I just created this PR in a case to learn how this works and maybe it could be useful to have it. That's this is under discussion at all :) :+1:

Reply:
It seems that it should broke update group announcements with such a change. For now you have one update group with 3 peers inside. If _subgroup_announce_check()_ return 0, then one update group will be split into two groups, where one is working as expected while another one - not.

Reply:
I'm interested what others think regarding this one :)

Reply:
It's an optimization you can not due with the current design without great difficulty and pain and breakages and significant performance degradation.